### ISSUE TITLE
Add Bulk Actions to Lists

### ISSUE THREAD

**Issue by mstmustisnt** [marmelab/admin-on-rest#65]
We can make items in list selectable.. But there is no way to manipulate those selected items.. It would be great to have an ability to add bulk actions component as well as filter.

Thank you for your awesome work!

**Comment by MattWilliamsDev** [comment]
+1

**Comment by wadjeroudi** [comment]
@fzaninotto Some ideas about the implementation of this feature ? Gonna try to implement it on a project.

**Comment by fzaninotto** [comment]
I'd advise to wait until material ui 1.0 is released. At that point we'll switch back from real tables to MUI tables.

**Comment by wadjeroudi** [comment]
@fzaninotto 
It seems that you switch backed to real tables with the responsive ui PR.
But you didn't reenabled the checkbox.
How do you imagine this feature. You check the boxes, and when you click a bulk action, you send all ids to the action ? Or Should we try to implement the rest call to get last status for all items selected and send objects to the action ?
What is the best way to get all checked boxes ?

**Comment by fzaninotto** [comment]
I haven't thought about it yet. Give it a try!

**Comment by mantis** [comment]
@fzaninotto @wadjeroudi - hi guys, how far have you got ?

I've been speaking to a friend and he was struggling to work out the best way to implement this.

Presumably, we'd have modify our UserList <List> component for instance to take an actions={<UserActions />}

I'm thinking then that bulk actions live at bottom of list (in datagrid) to distinguish between Add Filter/Create (non-bulk) actions at the top ?

Actions would then be a dropdown initially - a bit like filters work?

Once selected, pass in selectedRecords from the datagrid as a parameter.

I'm anticipating 2-3 types of actions:

a) action where it would immediately do a REST API post request
b) action that throws up a dialog box
c) presumably a 3rd type where action could modify page (e.g. as filters do)

**Comment by wadjeroudi** [comment]
@mantis didn't do it.
I think that for now, we should just aim one type of action a function that receives selected records. and it's up to the user to do whatever he wants ?

**Comment by fzaninotto** [comment]
I haven't given it time, but you can look at how [ng-admin](https://ng-admin-book.marmelab.com/doc/reference/View.html) handles it - it's probably a good inspiration.

**Comment by kodepareek** [comment]
Have just implemented this using Material UI components following are the steps I took. 

1) Create custom Datagrid using material UI (pretty straightforward) and basically copypasting AOR code and enabling checkboxes
2) adding Custon Action to the admin component. In my case this was an 'Assign All' button. That maps to the following action

```
export const AssignAllToSelf = (selection, basePath) => {
    return {
    type: ASSIGN_ALL_TO_SELF,
    payload: {selection: selection},
    meta: { resource: 'tales/assignEditor', fetch: UPDATE, cancelPrevious: false }
  }
};

```
3) adding a custom wrapper on the rest client that intercepts UPDATE calls to 'tales/assignEditor'
(AOR does not have bulk update feature just yet. )

```
if (type === 'UPDATE'  &&  resource === 'tales/assignEditor') {
    let url = config.host + '/' + 'tales/assignEditor'
    options.method = 'PUT';
    options.body = JSON.stringify(params)
    return handleRequestAndResponse(url, options)
  } else {
    return requestHandler(type, resource, params);
}


```

Working just fine! :)

**Comment by wadjeroudi** [comment]
@kodepareek how do you get the selection ? After you turn on the displayRowCheckbox, how do you link the custom(s) action(s) with the selected checkboxes ?

**Comment by kodepareek** [comment]
I write the selection into state as state.selection which is then picked up by the Assign Button action component. 

I have made the List into a connected component. You can also make the customData grid connected. 

```
class CustomDatagrid extends Component {
  handleRowClick = (rowArr) => {
    const {ids, handleSelection} = this.props
    if (Array.isArray(rowArr)) {
      handleSelection(rowArr.map(row => ids[row]))
    } else {
      handleSelection(ids)
    }
  }
  render() {
    const { resource, children, ids, isLoading, data, currentSort, basePath, styles = defaultStyles, muiTheme, rowStyle, options, headerOptions, bodyOptions, rowOptions } = this.props;
    return ( ...............

```


Inside the wrapping List component, in my case EditorAssignView

```
class EditorAssignView extends Component {
  handleSelection = (selection) => {
    const { AddSelectionToState } = this.props
    AddSelectionToState(selection)
  }
  render() {
    const props = this.props
    return (
      <List {...props} actions={<TaleAssignListActions />} title="Fresh Tales" >
         <CustomDatagrid handleSelection={this.handleSelection}>
     .............................................................

```

The add selection to state action

```
export const ADD_SELECTION_TO_STATE = 'ADD_SELECTION_TO_STATE';
export const AddSelectionToState = (selection, basePath) => {
    return {
    type: ADD_SELECTION_TO_STATE,
    payload: {selection: selection},
  }
};

```

**Comment by wadjeroudi** [comment]
Thx @kodepareek.
FYI, you don't need to rewrite the table part in your CustomDatagrid, you just can do that : 
```js
 <Datagrid headerOptions={{ adjustForCheckbox: true }} bodyOptions={{ displayRowCheckbox: true }} rowOptions={{ selectable: true }} options={{ multiSelectable: true, onRowSelection: {this.handleRowClick} }}>
```
It should be cleaner to maintain.

**Comment by kodepareek** [comment]
Oh That must be new. In my aor version selectable = false has been hard coded in.

**Comment by wadjeroudi** [comment]
It is harcoded but you can override them.

**Comment by kodepareek** [comment]
Oh that's super. Will try it out tomorrow. Will definitely make things cleaner.

**Comment by wadjeroudi** [comment]
@kodepareek sorry to bother, but when I click on my button, the selection is resetted immediatly because the checkboxes are unchecked if you click outside, do you have the problem ?

**Comment by alexeyoganezov** [comment]
What is the status of this feature? The project I'm currently working on requires bulk deletion, so I have to implement it in the nearest future. I'd like to know where to start: should I start from scratch or there is some existing project/codebase available that I could join or reuse?

**Comment by kodepareek** [comment]
Lists accept actions as props now. 

https://marmelab.com/admin-on-rest/List.html#actions

@djhi I think we can close this issue.

**Comment by fzaninotto** [comment]
No, actions (the list of buttons on the top of the screen) and list actions (actions you can execute on a selection of rows) are different. 

This feature is not started. If you want to give it a try, you're welcome!

**Comment by kodepareek** [comment]
I think I can do it. I will get back here a plan about how I am going to execute this feature. Have done something similar. Will think about generalizing.

**Comment by iamsimakov** [comment]
what news about feature?

**Comment by kodepareek** [comment]
Sorry guys. My app has been in some kind of production hell here at work. 

I should start work on this by this weekend.

**Comment by alexeyoganezov** [comment]
@kodepareek Could you, please, say what are you going to do? There is already a pull request with this feature: https://github.com/marmelab/admin-on-rest/pull/1013, or am I wrong? I'm waiting there for some feedback from... anyone.

**Comment by Toosick** [comment]
@kodepareek Did you use @alexeyoganezov changes or did you try a custom solution?

**Comment by kodepareek** [comment]
I made a custom solution. I think I have documented it somewhere on Github
or Stackoverflow. Will check and share.

**Comment by Toosick** [comment]
Oh I think it's in this thread correct?

**Comment by kodepareek** [comment]
Aah yes it right here above.

I big change would be that one does not need to create a custom DataGrid
component to make the checkboxes show. The rest is quite straightforward I
think.

**Comment by tkvw** [comment]
I created a PR for this: https://github.com/marmelab/admin-on-rest/pull/1241

**Comment by fzaninotto** [comment]
Implemented in #1521

**Comment by wujohns** [comment]
There no docs about how to edit selected items. I write the dataProvider....but where is the button(or how to add the bulk edit button)
